### PR TITLE
Support for 4 level deep delegated targets

### PR DESCRIPTION
**Pull request by kipz**
* Adds support for 4 levels of delegated role paths (related to #220) so that Rego policy bundle layouts can be stored and managed
* Top level targets remain unchanged (`*`)
* Generated delegate role metadata contains 4 paths (see below)
* Good metadata is used to scan for new files so that delegated targets without the 4 paths are unaffected by this change to aid rollout

Example generated delegated role `paths` metadata

```python
['myrole/*',
 'myrole/*/*',
 'myrole/*/*/*', 
 'myrole/*/*/*/*']
```

Edit:. I've manually tested the following scenarios:

* New files are not detected for old delegated role paths metadata that only contain one path (outside that path)
* Creating new roles creates delegated role metadata containing 4 paths (as above)
* New files inside the 4 paths are picked up and added to the signing event

### CODE REVIEW

**Review by jku — COMMENTED** [review]
I'll leave this initial review now as I think I need to think on this a bit more:
* I'm a little worried about trusting the metadata too much in signing event: A malicious signer can do all kinds of mischief but I would like to minimize it as much as possible
* so maybe when we're looking for change artifacts we don't just trust the paths from the metadata in the signing event, but instead only find artifacts in "hard coded" glob patterns -- or alternatively we must check that the patterns are a subset of what we support before we use them
* would be great to have a noisy failure if we find artifacts that are not in any defined patterns

Refactoring and combining the "find changed roles" and "find artifacts for role X" tasks might make the above easier to accomplish -- I'll have to re-read the code to be able to tell (it's been a while since I looked at it)

Anyway, I will try this out and have another read through this week, just wanted to write something down already ...

Once we are happy with this I can write an integration test for this

**Comment by jku** [comment]
> * so maybe when we're looking for changed artifacts we don't just trust the paths from the metadata in the signing event, but instead only find artifacts in "hard coded" glob patterns -- or alternatively we must check that the patterns are a subset of what we support before we use them

It's actually possible I'm too paranoid here: for delegated roles the patterns come from `targets`. It will be blindingly obvious in the signing event for a delegated role if `targets` metadata is being modified

File: repo/tuf_on_ci/_repository.py [hunk 1]
```diff
@@ -331,27 +333,29 @@ def _validate_role(
...
-    def _build_targets(target_dir: str, rolename: str) -> dict[str, TargetFile]:
+    def _build_targets(self, target_dir: str, rolename: str) -> dict[str, TargetFile]:
         """Build a roles dict of TargetFile based on target files in a directory"""
         targetfiles = {}
 
-        if rolename == "targets":
-            root_dir = target_dir
-        else:
-            root_dir = os.path.join(target_dir, rolename)
+        patterns = ["*"]
```
[the remark this answers is not in the document]
Comment (jku) [reply repo/tuf_on_ci/_repository.py:340]: maybe it's acceptable to approach it like this:

* for every targetpath PATH:
  * if PATH matches a delegated roles path pattern: add the artifact to that role
  * else if PATH is within accepted dir depth (so in this PR  I believe 5 levels), add artifact to "targets" role
  * else: PATH is too many levels deep, error 

this way you could still have multiple directory levels handled in top-level  "targets" but there is no misunderstanding about ownership of any artifacts: if the role ROLE exists, it owns everything under "ROLE/" directory.

**Comment by kipz** [comment]
> I'll leave this initial review now as I think I need to think on this a bit more:
> 
> * I'm a little worried about trusting the metadata too much in signing event: A malicious signer (or actually maintainer I suppose) can do all kinds of mischief but I would like to minimize it as much as possible
> * so maybe when we're looking for changed artifacts we don't just trust the paths from the metadata in the signing event, but instead only find artifacts in "hard coded" glob patterns -- or alternatively we must check that the patterns are a subset of what we support before we use them
> * would be great to have a noisy failure if we find artifacts that are not in any defined patterns
> 
> Refactoring and combining the "find changed roles" and "find artifacts for role X" tasks might make the above easier to accomplish -- I'll have to re-read the code to be able to tell (it's been a while since I looked at it)
> 
> Anyway, I will try this out and have another read through this week, just wanted to write something down already ...
> 
> Once we are happy with this I can write an integration test for this

Yeah, I did consider this. In fact, my first implementation enforced the same patterns for generation and scanning, but due to the project layout (as far as I can tell as relative python noob), there's no way to share code/config between the modules, and I didn't want to open up that can of worms for this change. I just duplicated the configuration and algorithm between the two projects.

File: repo/tuf_on_ci/_repository.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kipz) [reply repo/tuf_on_ci/_repository.py:340]: My original PR #220 handled things in this way, and I do think it works, but the it gets complicated when new roles are added, do files previously owned by targets now get picked up? Similarly with removing a role, do those files now get picked up by targets?

[the remark this answers is not in the document]
Comment (jku) [reply repo/tuf_on_ci/_repository.py:340]: yeah that would be the implication:
* it's true that this is more complicated than just not allowing top level targets to own artifacts in subdirectories
* but I think it's manageable if delegation path pattern is now not a free-for-all: every artifact is owned by either the logical delegated role or top level targets role 

but I am fine with not making this PR more complicated

**Review by jku — COMMENTED** [review]
I've played with this now (sorry it took a while). I think it's good.

Some notes:
* `update_targets()` and `_find_changed_target_roles()` are clearly asking for a refactor but I don't have concrete suggestions so we can leave that for future
* signing-event does not fail if there are artifacts deeper than supported level -- this is not any different from current situation so fine by me: should file a bug though
* top-level targets still does not support multiple levels (in the case where the delegation does not exist): I think it should but since this is also not currently supported, seems fine to not implement it here
* There is no migration for existing roles:  new roles in existing repos will handle deep delegated paths but existing roles will not... that's a bit confusing but I don't really know what else would make sense

TODO:
* I'll try to come up with an integration test now
* We should document this. docs/SIGNER-MANUAL.md seems to be the only place talking about artifacts


@kommendorkapten can you have a look, any comments? See https://github.com/jku/tuf-on-ci-sigstore-test/pull/52 for a test signing event

**Comment by kommendorkapten** [comment]
@jku of course, busy today but I'll get on it first thing in the morning 👍

**Comment by jku** [comment]
To reproduce CI failures locally: 
* You can run linter for both signer and repo with `tox -m lint`
* Likewise `tox -m test`  runs all the tests (these should pass now)

File: repo/test/test_ci_repository.py [hunk 2]
```diff
@@ -41,6 +45,77 @@ def test_default_signing_days(self):
...
+            targets = repo.targets("targets")
+            self.assertNotIn("tfile1.txt", targets.targets)
+
+            # adding these files and updating adds them to targets
+            shutil.copy(os.path.join(src_targets, "tfile1.txt"), temp_targets)
+            repo.update_targets("targets")
+            targets = repo.targets("targets")
+            self.assertIn("tfile1.txt", targets.targets)
+
+
```
Comment (kommendorkapten) [inline repo/test/test_ci_repository.py:77]: Nit: two new lines here

File: repo/test/test_ci_repository.py [hunk 3]
```diff
@@ -41,6 +45,77 @@ def test_default_signing_days(self):
...
+            # now add a file not covered by another role
+            shutil.copy(os.path.join(src_targets, "tfile1.txt"), temp_targets)
```
Comment (kommendorkapten) [inline repo/test/test_ci_repository.py:79]: Isn't this the exact same file that's being added on line 72?

**Comment by kommendorkapten** [comment]
Code looks good in general, some comments on tests only.

**Comment by jku** [comment]
I'll take a look at the e2e test failure: maybe I forgot to commit some expected test result

**Review by jku — APPROVED** [review]
lgtm, thanks

**Review by kommendorkapten — APPROVED** [review]
Thanks

**Pull request merged**